AsyncDisplayKit/ASComponentHostingNode.mm

```diff
@@ -0,0 +1,216 @@
+  }];
+  [self addSubnode:_containerNode];
+
+  _componentProvider = componentProvider;
+  _sizeRangeProvider = sizeRangeProvider;
+
+  _pendingInputs = { .scopeRoot = [CKComponentScopeRoot rootWithListener:self] };
+  _componentNeedsUpdate = YES;
+  _requestedUpdateMode = CKUpdateModeAsynchronous;
+
+  [self _asynchronouslyUpdateComponentIfNeeded];
+
+  return self;
+}
+
+- (void)dealloc
+{
+  CKUnmountComponents(_mountedComponents);
+}
+
+#pragma mark - ASDisplayNode+Subclasses
+
+- (void)layout
+{
+  [super layout];
+
+  _containerNode.frame = self.bounds;
+  const CGSize size = self.calculatedSize;
+  if (_mountedLayout.component != _component || !CGSizeEqualToSize(_mountedLayout.size, size)) {
+    _mountedLayout = CKComputeComponentLayout(_component, {size, size}, size);
+  }
+  _mountedComponents = CKMountComponentLayout(_mountedLayout, _containerNode.view, _mountedComponents, nil);
+}
+
+- (CGSize)calculateSizeThatFits:(CGSize)constrainedSize
+{
+  const CKSizeRange componentConstrainedSize = [_sizeRangeProvider sizeRangeForBoundingSize:constrainedSize];
+  return CKComputeComponentLayout(_component, componentConstrainedSize, componentConstrainedSize.max).size;
+}
+
+#pragma mark - CKComponentStateListener
+
+- (void)componentScopeHandleWithIdentifier:(CKComponentScopeHandleIdentifier)globalIdentifier
+                            rootIdentifier:(CKComponentScopeRootIdentifier)rootIdentifier
+                     didReceiveStateUpdate:(nullable id (^)(_Nullable id))stateUpdate
+                                      mode:(CKUpdateMode)mode
+{
+  _pendingInputs.stateUpdates.insert({globalIdentifier, stateUpdate});
+  [self _setNeedsUpdateWithMode:mode];
+}
+
+#pragma mark - Accessors
+
+- (void)updateModel:(nullable id<NSObject>)model mode:(CKUpdateMode)mode
+{
+  _pendingInputs.model = model;
+  [self _setNeedsUpdateWithMode:mode];
+}
+
+- (void)updateContext:(nullable id<NSObject>)context mode:(CKUpdateMode)mode
+{
+  _pendingInputs.context = context;
+  [self _setNeedsUpdateWithMode:mode];
+}
+
+#pragma mark - Private
+
+- (void)_setNeedsUpdateWithMode:(CKUpdateMode)mode
+{
+  if (_componentNeedsUpdate && _requestedUpdateMode == CKUpdateModeSynchronous) {
+    return; // Already scheduled a synchronous update; nothing more to do.
+  }
+
+  _componentNeedsUpdate = YES;
+  _requestedUpdateMode = mode;
+
+  switch (mode) {
+    case CKUpdateModeAsynchronous:
+      [self _asynchronouslyUpdateComponentIfNeeded];
+      break;
+    case CKUpdateModeSynchronous:
+      [self _synchronouslyUpdateComponentIfNeeded];
+      [self _invalidateLayout];
+      break;
+  }
+}
+
+- (void)_asynchronouslyUpdateComponentIfNeeded
+{
+  if (_scheduledAsynchronousComponentUpdate) {
+    return;
+  }
+  _scheduledAsynchronousComponentUpdate = YES;
+
+  // Wait until the end of the run loop so that if multiple async updates are triggered we don't thrash.
+  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
```

Comment:
@ocrickard: 👍

How about moving to an `NSOperationQueue` with a cap on concurrency? Or just a single serial dispatch queue that's off the main thread?